pkg/backup/util/util.go

```diff
@@ -123,83 +134,122 @@ func GenerateGcsCertEnvVar(secret *corev1.Secret, gcs *v1alpha1.GcsStorageProvid
 	var reason string
 
-	switch backup.Spec.StorageType {
+	switch storageType {
 	case v1alpha1.BackupStorageTypeS3:
-		s3SecretName := backup.Spec.S3.SecretName
+		if provider.S3 == nil {
+			return certEnv, "S3ConfigIsEmpty", errors.New("s3 config is empty")
+		}
+		s3SecretName := provider.S3.SecretName
 		secret, err := secretLister.Secrets(ns).Get(s3SecretName)
 		if err != nil {
-			err := fmt.Errorf("backup %s/%s get s3 secret %s failed, err: %v", ns, name, s3SecretName, err)
+			err := fmt.Errorf("get s3 secret %s/%s failed, err: %v", ns, s3SecretName, err)
 			return certEnv, "GetS3SecretFailed", err
 		}
 
 		keyStr, exist := CheckAllKeysExistInSecret(secret, constants.S3AccessKey, constants.S3SecretKey)
 		if !exist {
-			err := fmt.Errorf("backup %s/%s, The s3 secret %s missing some keys %s", ns, name, s3SecretName, keyStr)
+			err := fmt.Errorf("s3 secret %s/%s missing some keys %s", ns, s3SecretName, keyStr)
 			return certEnv, "s3KeyNotExist", err
 		}
 
-		certEnv, reason, err = GenerateS3CertEnvVar(secret, backup.Spec.S3.DeepCopy())
+		certEnv, reason, err = GenerateS3CertEnvVar(provider.S3.DeepCopy())
 		if err != nil {
 			return certEnv, reason, err
 		}
 	case v1alpha1.BackupStorageTypeGcs:
-		gcsSecretName := backup.Spec.Gcs.SecretName
+		if provider.Gcs == nil {
+			return certEnv, "GcsConfigIsEmpty", errors.New("gcs config is empty")
+		}
+		gcsSecretName := provider.Gcs.SecretName
 		secret, err := secretLister.Secrets(ns).Get(gcsSecretName)
 		if err != nil {
-			err := fmt.Errorf("backup %s/%s get gcs secret %s failed, err: %v", ns, name, gcsSecretName, err)
+			err := fmt.Errorf("get gcs secret %s/%s failed, err: %v", ns, gcsSecretName, err)
 			return certEnv, "GetGcsSecretFailed", err
 		}
 
 		keyStr, exist := CheckAllKeysExistInSecret(secret, constants.GcsCredentialsKey)
 		if !exist {
-			err := fmt.Errorf("backup %s/%s, The gcs secret %s missing some keys %s", ns, name, gcsSecretName, keyStr)
+			err := fmt.Errorf("the gcs secret %s/%s missing some keys %s", ns, gcsSecretName, keyStr)
 			return certEnv, "gcsKeyNotExist", err
 		}
 
-		certEnv, reason, err = GenerateGcsCertEnvVar(secret, backup.Spec.Gcs)
+		certEnv, reason, err = GenerateGcsCertEnvVar(provider.Gcs)
 
 		if err != nil {
 			return certEnv, reason, err
 		}
 	default:
-		err := fmt.Errorf("backup %s/%s don't support storage type %s", ns, name, backup.Spec.StorageType)
+		err := fmt.Errorf("don't support storage type %s", storageType)
 		return certEnv, "NotSupportStorageType", err
 	}
 	return certEnv, reason, nil
 }
 
-// GetTidbUserAndPassword get the tidb user and password from specific secret
-func GetTidbUserAndPassword(ns, name, tidbSecretName string, secretLister corelisters.SecretLister) (user, password, reason string, err error) {
+// GenerateTidbPasswordEnv generate the password EnvVar
+func GenerateTidbPasswordEnv(ns, name, tidbSecretName string, secretLister corelisters.SecretLister) ([]corev1.EnvVar, string, error) {
```

Comment:
I think we can follow k8s builtin controller's behavior to not checking the referenced resources existence.

Reply:
The purpose of the check is to make it fail quickly when there is a problem. At the same time, the failure information and reason will be written into the backup CR status, which is helpful for upper-layer application integrated with us to diagnose the problem quickly. k8s only discovers problems when it encounters problems in the execution layer, and then records the problem to the underlying pod object, which is not very straightforward for diagnosing the problem of the upper-layer application